aws-lc-rs/src/aead/aead_ctx.rs

```diff
@@ -0,0 +1,100 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR ISC
+
+use crate::aead::TAG_LEN;
+use crate::cipher::chacha;
+
+use crate::cipher::aes::{AES_128_KEY_LEN, AES_256_KEY_LEN};
+use crate::error::Unspecified;
+use crate::ptr::ConstPointer;
+use aws_lc::{
+    EVP_AEAD_CTX_cleanup, EVP_AEAD_CTX_init, EVP_AEAD_CTX_zero, EVP_aead_aes_128_gcm,
+    EVP_aead_aes_256_gcm, EVP_aead_chacha20_poly1305, EVP_AEAD, EVP_AEAD_CTX,
+};
+use std::mem::MaybeUninit;
+use std::ptr::null_mut;
+
+#[allow(
+    clippy::large_enum_variant,
+    variant_size_differences,
+    non_camel_case_types
+)]
+pub(crate) enum AeadCtx {
+    AES_128_GCM(EVP_AEAD_CTX),
+    AES_256_GCM(EVP_AEAD_CTX),
+    CHACHA20_POLY1305(EVP_AEAD_CTX),
+}
+
+unsafe impl Send for AeadCtx {}
+unsafe impl Sync for AeadCtx {}
+
+impl AeadCtx {
+    pub(crate) fn aes_128_gcm(key_bytes: &[u8]) -> Result<Self, Unspecified> {
+        if AES_128_KEY_LEN != key_bytes.len() {
+            return Err(Unspecified);
+        }
+        unsafe {
+            let aead = ConstPointer::new(EVP_aead_aes_128_gcm())?;
+            Ok(AeadCtx::AES_128_GCM(AeadCtx::build_context(
+                &aead, key_bytes,
+            )?))
+        }
+    }
+
+    pub(crate) fn aes_256_gcm(key_bytes: &[u8]) -> Result<Self, Unspecified> {
+        if AES_256_KEY_LEN != key_bytes.len() {
+            return Err(Unspecified);
+        }
+        unsafe {
+            let aead = ConstPointer::new(EVP_aead_aes_256_gcm())?;
+            Ok(AeadCtx::AES_256_GCM(AeadCtx::build_context(
+                &aead, key_bytes,
+            )?))
+        }
+    }
+
+    pub(crate) fn chacha20(key_bytes: &[u8]) -> Result<Self, Unspecified> {
+        if chacha::KEY_LEN != key_bytes.len() {
+            return Err(Unspecified);
+        }
+        unsafe {
+            let aead = ConstPointer::new(EVP_aead_chacha20_poly1305())?;
+            Ok(AeadCtx::CHACHA20_POLY1305(AeadCtx::build_context(
+                &aead, key_bytes,
+            )?))
+        }
+    }
+
+    unsafe fn build_context(
```

Comment:
If you isolate the `unsafe` into the body, you could remove needing to wrap all the functions above.

Reply:
The call to get the `*const EVP_AEAD` (e.g., `EVP_aead_aes_128_gcm()`) also requires an unsafe block. I don't see a good way of eliminating the unsafe block from the calling function.

Reply:
I was able to avoid several unsafe blocks by passing a function pointer to `build_context`.